services/s3/src/main/java/software/amazon/awssdk/services/s3/S3Utilities.java

```diff
@@ -0,0 +1,231 @@
+ */
+
+package software.amazon.awssdk.services.s3;
+
+import java.net.MalformedURLException;
+import java.net.URI;
+import java.net.URL;
+import java.util.Optional;
+import software.amazon.awssdk.annotations.Immutable;
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.auth.signer.AwsSignerExecutionAttribute;
+import software.amazon.awssdk.awscore.AwsExecutionAttribute;
+import software.amazon.awssdk.awscore.client.config.AwsClientOption;
+import software.amazon.awssdk.awscore.internal.EndpointUtils;
+import software.amazon.awssdk.core.SdkRequest;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.client.config.SdkClientConfiguration;
+import software.amazon.awssdk.core.client.config.SdkClientOption;
+import software.amazon.awssdk.core.interceptor.Context;
+import software.amazon.awssdk.core.interceptor.ExecutionAttributes;
+import software.amazon.awssdk.core.sync.RequestBody;
+import software.amazon.awssdk.http.SdkHttpFullRequest;
+import software.amazon.awssdk.http.SdkHttpMethod;
+import software.amazon.awssdk.http.SdkHttpRequest;
+import software.amazon.awssdk.protocols.core.OperationInfo;
+import software.amazon.awssdk.protocols.core.PathMarshaller;
+import software.amazon.awssdk.protocols.core.ProtocolUtils;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.s3.internal.handlers.EndpointAddressInterceptor;
+import software.amazon.awssdk.services.s3.model.GetUrlRequest;
+import software.amazon.awssdk.utils.Validate;
+
+/**
+ * Utilities for working with Amazon S3 objects. An instance of this class can be created by:
+ * 1) Directly using the {@link #builder()} method. You have to manually specify the configuration params like region,
+ * s3Configuration on the builder.
+ * 2) Using the low-level client {@link S3Client#utilities()} method. This is recommended as SDK will use the same
+ * configuration from the {@link S3Client} object to create the {@link S3Utilities} object.
+ *
+ * Note: This class should not have methods that make network calls.
+ */
+@Immutable
+@SdkPublicApi
+public final class S3Utilities {
+
+    private final Region region;
+
+    private final S3Configuration s3Configuration;
+
+    /**
+     * SDK currently validates that region is present while constructing {@link S3Utilities} object.
+     * This can be relaxed in future when more methods are added that doesn't use region.
+     */
+    public S3Utilities(Builder builder) {
+        this.region = Validate.paramNotNull(builder.region, "Region");
+        this.s3Configuration = builder.s3Configuration;
+    }
+
+    /**
+     * Creates a builder for {@link S3Utilities}.
+     */
+    public static Builder builder() {
+        return new Builder();
+    }
+
+    // Used by low-level client
+    @SdkInternalApi
+    static S3Utilities create(SdkClientConfiguration clientConfiguration) {
```

Comment:
Should we have a zero-arg create() method that uses the default client configuration? We could even cheat a bit and do...

```Java
return S3Client.builder().httpClient(NoopHttpClient.create()).build().utilities();
```

Reply:
`create` method is for internal use. I prefer to avoid direct defaults in the S3Utilities class and let customer configure the fields they want or use the S3Client utilities method.